### ISSUE TITLE
[FEATURE REQUEST] Run command before updating.

### ISSUE THREAD

**Issue by Goth667** [marticliment/UniGetUI#3038]
### Please confirm these before moving forward.

- [X] I have searched for my feature proposal and have not found a work-in-progress/duplicate/resolved/discarded issue.
- [X] This proposal is a completely new feature. If you want to suggest an improvement or an enhancement, please use [this template](https://github.com/marticliment/WingetUI/issues/new?assignees=marticliment&labels=enhancement&projects=&template=enhancement-improvement.yml&title=%5BENHANCEMENT%5D+%28Enter+your+description+here%29).

### Describe the new feature

Hi,
I want do propose a feature in the installation options of a package. It would be nice if it was possible to run a command before the installation or update. 
A simple text field where you can add the command you want executed and a checkbox if it should run as Administrator. 
Thanks

### Describe how this new feature could help users

Some packages can't automatically be installed if the program is already running, hanging up or its files are already in use by another application with a shared library. Depending on the system, it would be possible to add a command that shuts down the application or deletes the settings file etc.
![image](https://github.com/user-attachments/assets/d04292bc-9bd2-4dee-b64d-05e1644fcef0)

**Comment by mrixner** [comment]
Duplicate of #2160 and #2193

**Comment by Goth667** [comment]
@mrixner 
Sry, apparently I did a bad job a searching before posting.

@marticliment 
Would adding a user for impersonation to run the command work?

**Comment by Goth667** [comment]
Thanks for the explanation.

So having to set up a user under whom the command will be executed would be harder to do then to double click a .reg file. 
Should we also limit it so the impersonated user can't be the current user?
A warning pop up would also help, I guess.

In the end if you get the user to run a .reg or script there is no need to do it in UniGetUI.

In the end it boils down to the war against the universe. Where as devs try to make software safer and safer for dumb users and the universe trying to produce dumber and dumber users.

**Comment by mrixner** [comment]
One method I could think of is simply asking the user for confirmation about every command; every time it tries to run a post-install command, UniGetUI will check if that _specific command_ has been approved before, and if not throw up an approval dialog. This way, an externally modified command will need approval first. And to prevent the external command from just being set as approved, maybe there could be a "hook password" in the settings used to encrypt the list of approved commands?

**Comment by marticliment** [comment]
> UniGetUI will check if that _specific command_ has been approved before, and if not throw up an approval dialog.

Yet again, any information that is stored on the PC could have already been compromised, since it is not stored under safe zones. I reckon that, however, if post-install commands are only run when UniGetUI is NOT run as administrator, there would be no possible exploit, since UniGetUI would execute your script in local user rights, but you already needed local user rights to force UniGetUI into performing such actions

**Comment by mrixner** [comment]
> Yet again, any information that is stored on the PC could have already been compromised, since it is not stored under safe zones.

Even an one-way encrypted (I forgot the word for this, but this was what I meant instead of encryption) password? I imagine the password wouldn't just be stored in plaintext.

**Comment by Goth667** [comment]
> > UniGetUI will check if that _specific command_ has been approved before, and if not throw up an approval dialog.
> > there could be a "hook password" in the settings
> 
> Yet again, any information that is stored on the PC could have already been compromised, since it is not stored under safe zones. I reckon that, however, if post-install commands are only run when UniGetUI is NOT run as administrator, there would be no possible exploit, since UniGetUI would execute your script in local user rights, but you already needed local user rights to force UniGetUI into performing such actions

As for the command before, would it be possible to kill the process of the application that will be installed with the package?

**Comment by marticliment** [comment]
> Even an one-way encrypted (I forgot the word for this, but this was what I meant instead of encryption) password? I imagine the password wouldn't just be stored in plaintext.

You mean hashing? How would it decrypt the contents of the file?




> As for the command before, would it be possible to kill the process of the application that will be installed with the package?

What do you mean?

**Comment by mrixner** [comment]
> You mean hashing? How would it decrypt the contents of the file?

Yeah, hashing! It wouldn't need to decrypt anything; just hash whatever command is about to be run and if it's hash is listed as an approved command then the command could be run, otherwise prompt the user about the command. This is how account passwords are implemented, is it not?

**Comment by marticliment** [comment]
> This is how account passwords are implemented, is it not?

More or less yes, but still wouldn't solve the issue, since a malicious agent could still calculate the hash of their command and inject it into the database where it is stored

**Comment by mrixner** [comment]
> More or less yes, but still wouldn't solve the issue, since a malicious agent could still calculate the hash of their command and inject it into the database where it is stored

That's where the password come in. You can hash the hook password normally and compare against the saved hash to see if it's correct; the hook password is only known by the user, so it doesn't need a special hashing algorithm. Then use a hashing algorithm that takes an input (a password / secret key) to hash the command, so the attacker cannot get the hash of the command without the user's hook password.

**Comment by marticliment** [comment]
Huh, this looks interesting. I will investigate if this would be a viable solution.

**Comment by Goth667** [comment]
Thank you both for putting the time into this. I really appreciate it.

**Comment by the-sloth** [comment]
Any updates on this feature request regarding the implementation of Pre/Post install hooks?

**Comment by mrixner** [comment]
I would be happy to implement the details if Martí finds a way to mitigate the security risks.

**Comment by marticliment** [comment]
Hello,

I have published a PreRelease version of UniGetUI that allows to specify pre/post install/update/uninstall commands/scripts. I'd appreciate if you could test it and provide some feedback.

Thanks!
Download link: [3.2.1-beta1 (release)](https://github.com/marticliment/UniGetUI/releases/tag/3.2.1-beta1)

**Comment by Goth667** [comment]
@marticliment wow thanks a lot, just tested it with a special needs program and worked like a charm.
